dll/cpl/input/settings_page.c

```diff
@@ -283,15 +426,32 @@ OnCommandSettingsPage(HWND hwndDlg, WPARAM wParam)
             HWND hwndList = GetDlgItem(hwndDlg, IDC_KEYLAYOUT_LIST);
             if (hwndList)
             {
-                LVITEM item = { LVIF_PARAM };
-                item.iItem = ListView_GetNextItem(hwndList, -1, LVNI_SELECTED);
+                HTREEITEM hItem = TreeView_GetSelection(hwndList);
+                TV_ITEM item = { TVIF_HANDLE | TVIF_PARAM };
+                item.hItem = hItem;
 
-                if (ListView_GetItem(hwndList, &item))
+                if (hItem && TreeView_GetItem(hwndList, &item))
                 {
-                    InputList_Remove((INPUT_LIST_NODE*) item.lParam);
-                    UpdateInputListView(hwndList);
-                    SetControlsState(hwndDlg);
-                    PropSheet_Changed(GetParent(hwndDlg), hwndDlg);
+                    if (item.lParam == 0) // Keyboard item?
+                    {
+                        item.hItem = hItem = TreeView_GetParent(hwndList, hItem);
+                        TreeView_GetItem(hwndList, &item);
+                    }
+
+                    if (HIWORD(item.lParam)) // Leaf?
+                    {
+                        InputList_Remove((INPUT_LIST_NODE*) item.lParam);
+                        UpdateInputListView(hwndList);
+                        SetControlsState(hwndDlg);
+                        PropSheet_Changed(GetParent(hwndDlg), hwndDlg);
+                    }
+                    else // Root?
+                    {
+                        InputList_RemoveByLang(LOWORD(item.lParam));
+                        UpdateInputListView(hwndList);
+                        SetControlsState(hwndDlg);
+                        PropSheet_Changed(GetParent(hwndDlg), hwndDlg);
+                    }
```

Comment:
It seems this can be simplified to:
```suggestion
                    if (HIWORD(item.lParam)) // Leaf?
                        InputList_Remove((INPUT_LIST_NODE*)item.lParam);
                    else // Root?
                        InputList_RemoveByLang(LOWORD(item.lParam));

                    UpdateInputListView(hwndList);
                    SetControlsState(hwndDlg);
                    PropSheet_Changed(GetParent(hwndDlg), hwndDlg);
```

Reply:
Fixed.